meta/src/main/scala/meta/RisePrimitiveGenerator.scala

```diff
@@ -0,0 +1,268 @@
+package meta
+
+import fastparse.{Parsed, parse}
+import meta.NatParser.NatAST
+import meta.TypeParser.TypeAST
+import meta.TypeParser.TypeAST.{FragmentAST, MatrixLayoutAST}
+
+object RisePrimitiveGenerator {
+  def main(args: Array[String]): Unit = {
+    val sourceDir = args.head
+    val rise = os.Path(sourceDir) / "rise"
+    os.walk.stream(rise).filter(_.ext == "rise").foreach(path => {
+
+      val definition = os.read(path)
+      parse(definition, TypeParser.PrimitiveDeclarations(_)) match {
+        case failure: Parsed.Failure =>
+          println(s"Failed to parse `${failure.extra.input}'")
+          println(s"  $failure")
+        case Parsed.Success(seq, _) =>
+          seq.foreach {
+            case (name, args, typeSignature) if TypeParser.isWellKindedType(typeSignature) =>
+              val outputPath = (path / os.up) / s"$name.scala"
+              println(s"Generate $outputPath")
+
+              val generatedDef = args match {
+                case None =>
+                  generateObject(name, typeSignature)
+                case Some((start, end)) =>
+                  generateCaseClass(name, definition.substring(start, end), typeSignature)
+              }
+
+              import scala.meta._
+              val packageName = path.relativeTo(rise).segments.dropRight(1).foldLeft[Term.Ref](Term.Name("rise")) {
+                case (t, name) => Term.Select(t, Term.Name(name))
+              }
+              val code = s"""// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! //
+                            |// This file is automatically generated and should not be changed manually //
+                            |// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! //
+                            |${q"""
+package $packageName {
+
+import rise.core.DSL._
+import rise.core.DSL.Type._
+import rise.core._
+import rise.core.types._
+import arithexpr.arithmetic._
+
+..${generatedDef.stats}
+
+}""".toString()}
+                            |""".stripMargin
+
+              os.write.over(outputPath, code)
+            case (name, _, typeSignature) =>
+              println(s"Could not generate code for `$name' as type signature `$typeSignature' is not well kinded.")
+          }
+      }
+    })
+  }
+
+  def generateObject(name: String, typeSignature: TypeAST): scala.meta.Term.Block = {
+    import scala.meta._
+    val generated = q"""{
+      object ${Term.Name{name}} extends Builder {
+        private final case class Primitive()(override val t: Type = TypePlaceholder)
+          extends rise.core.Primitive
+        {
+          override val name: String = ${Lit.String(name)}
+          override def setType(ty: Type): Primitive = Primitive()(ty)
+          override def typeScheme: Type = ${generateTypeScheme(typeSignature, Map.empty)}
+
+          override def equals(obj: Any): Boolean = obj match {
+            case p: Primitive => p.t =~~= t
```

Comment:
Should the default `equals` not be kept to be consistent with the `equals` of other language constructs?
https://github.com/rise-lang/shine/blob/5976011b902c1396acc145f2e66a4ff4a1fddfbd/src/main/scala/rise/core/Expr.scala#L28-L31

Reply:
Alpha equivalence now considers two primitives to be the same if their names are the same, I think it's safe to do the same here.

Reply:
I am doing exactly what the current macro in `master` is doing: https://github.com/rise-lang/shine/blob/master/macros/src/main/scala/rise/macros/Primitive.scala#L62

Reply:
We could leave things as they are here and open a separate issue?

Reply:
@umazalakain could you do this please and we merge this PR?

Reply:
Done: #141